File: rust/src/moveinfo.rs
//! Move-specific mechanics, transcribed from `src/pokeuraou/moveinfo.py`.
//!
//! Variable base power, a move's own type change, and the two effectiveness overrides.
//! `base_power` returns `None` for a move it cannot determine, exactly as the Python does:
//! the caller reports that rather than printing a number with no basis.

use crate::battler::Battler;
use crate::id::Id;
use crate::position::Types;
use crate::reg::Reg;

/// Weather and terrain are inline ids, as on the field they are copied from: as `String`s
/// they were two allocations per hit and per damage call without one (IKA-101).
#[derive(Debug, Clone, Default, serde::Deserialize)]
pub struct MoveContext {
    pub weather: Option<Id>,
    pub terrain: Option<Id>,
    pub side_total_fainted: i64,
    pub times_attacked: i64,
    pub target_hurt_this_turn: bool,
    pub damaged_by_target: bool,
    pub previous_move_failed: bool,
    pub hit_index: i64,
    pub moving_last: bool,
    pub ally_used_same_move: bool,
    /// What a reply's `damageCallback` returns this turn (`damage_callback::damage`), 0 when
    /// there is no turn to read it from (IKA-213).
    #[serde(default)]
    pub reply_damage: i64,
}

impl MoveContext {
    #[inline]
    pub fn weather_name(&self) -> Option<&str> {
        self.weather.as_ref().map(|w| w.as_str())
    }

    #[inline]
    pub fn terrain_name(&self) -> Option<&str> {
        self.terrain.as_ref().map(|t| t.as_str())
    }
}

const WEIGHT_BP: [(f64, i64); 5] =
    [(200.0, 120), (100.0, 100), (50.0, 80), (25.0, 60), (10.0, 40)];
const RELATIVE_WEIGHT_BP: [(f64, i64); 4] = [(5.0, 120), (4.0, 100), (3.0, 80), (2.0, 60)];

fn weather_ball_type(weather: Option<&str>) -> Option<&'static str> {
    match weather {
        Some("sunnyday") | Some("desolateland") => Some("Fire"),
        Some("raindance") | Some("primordialsea") => Some("Water"),
        Some("sandstorm") => Some("Rock"),
        Some("hail") | Some("snowscape") | Some("snow") => Some("Ice"),
        _ => None,
    }
}

fn terrain_pulse_type(terrain: Option<&str>) -> Option<&'static str> {
    match terrain {
        Some("electricterrain") => Some("Electric"),
        Some("grassyterrain") => Some("Grass"),
        Some("mistyterrain") => Some("Fairy"),
        Some("psychicterrain") => Some("Psychic"),
        _ => None,
    }
}

fn solar_weak(weather: Option<&str>) -> bool {
    matches!(
        weather,
        Some("raindance")
            | Some("primordialsea")
            | Some("sandstorm")
            | Some("hail")
            | Some("snowscape")
            | Some("snow")
    )
}

fn weight_bp(kg: f64) -> i64 {
    for (threshold, bp) in WEIGHT_BP {
        if kg >= threshold {
            return bp;
        }
    }
    20
}

fn relative_weight_bp(attacker_kg: f64, target_kg: f64) -> i64 {
    if target_kg <= 0.0 {
        return 120;
    }
    for (ratio, bp) in RELATIVE_WEIGHT_BP {
        if attacker_kg >= target_kg * ratio {
            return bp;
        }
    }
    40
}

pub fn fixed_damage(
    move_id: &str,
    attacker: &Battler,
    defender: &Battler,
    ctx: &MoveContext,
) -> Option<i64> {
    match move_id {
        "superfang" | "naturesmadness" | "ruination" => Some((defender.hp / 2).max(1)),
        "finalgambit" => Some(attacker.hp.max(1)),
        "endeavor" => Some((defender.hp - attacker.hp).max(0)),
        // Counter, Mirror Coat, Metal Burst, Comeuppance (IKA-213). Without a turn's record
        // (a scorer's context) the damage is unknown, as before.
        "counter" | "mirrorcoat" | "metalburst" | "comeuppance" if ctx.reply_damage > 0 => {
            Some(ctx.reply_damage)
        }
        _ => None,
    }
}

/// A move's own `onModifyType`, or None when the declared type stands.
pub fn effective_type(
    move_id: &str,
    attacker: &Battler,
    ctx: &MoveContext,
) -> Option<&'static str> {
    match move_id {
        "weatherball" => weather_ball_type(ctx.weather_name()),
        "terrainpulse" => terrain_pulse_type(crate::airborne::terrain_under(attacker, ctx)),
        "aurawheel" => {
            if attacker.species.as_str() == "morpekohangry" {
                Some("Dark")
            } else {
                None
            }
        }
        "ragingbull" => match attacker.species.as_str() {
            "taurospaldeacombat" => Some("Fighting"),
            "taurospaldeablaze" => Some("Fire"),
            "taurospaldeaaqua" => Some("Water"),
            _ => None,
        },
        _ => None,
    }
}

/// A move's `onEffectiveness`, as (multiplier, Showdown's typeMod).
pub fn effectiveness_override(
    reg: &Reg,
    move_id: &str,
    move_type: &str,
    defender_types: &Types,
) -> Option<(f64, i64)> {
    let step = |mult: f64| -> i64 {
        if mult == 2.0 {
            1
        } else if mult == 0.5 {
            -1
        } else {
            0
        }
    };
    match move_id {
        "freezedry" => {
            let row = &reg.type_chart[reg.type_slot_of(move_type)];
            let mut mult = 1.0;
            let mut steps = 0i64;
            for t in defender_types.as_slice() {
                if t.as_str() == "Water" {
                    mult *= 2.0;
                    steps += 1;
                    continue;
                }
                let value = row[reg.type_slot_of(t.as_str())];
                mult *= value;
                steps += step(value);
            }
            Some((mult, steps.clamp(-6, 6)))
        }
        "flyingpress" => {
            let mut mult = 1.0;
            let mut steps = 0i64;
            for t in defender_types.as_slice() {
                for name in ["Fighting", "Flying"] {
                    let value =
                        reg.type_chart[reg.type_slot_of(name)][reg.type_slot_of(t.as_str())];
                    mult *= value;
                    steps += step(value);
                }
            }
            Some((mult, steps.clamp(-6, 6)))
        }
        _ => None,
    }
}

pub fn is_approximate(move_id: &str) -> bool {
    matches!(
        move_id,
        "beatup"
            | "ficklebeam"
            | "lashout"
            | "mistyexplosion"
            | "gravapple"
            | "shellsidearm"
            | "boltbeak"
            | "fishiousrend"
            | "punishment"
            | "trumpcard"
            | "wringout"
            | "crushgrip"
            | "spitup"
            | "naturalgift"
            | "fling"
            | "brine"
            | "retaliate"
            | "smellingsalts"
            | "wakeupslap"
            | "watershuriken"
    )
}

/// Base power for one hit, or None when this module cannot determine it.
pub fn base_power(
    reg: &Reg,
    move_id: &str,
    attacker: &Battler,
    defender: &Battler,
    ctx: &MoveContext,
) -> Option<i64> {
    let mv = reg.moves.get(move_id)?;
    let declared = mv.base_power;

    match move_id {
        "lowkick" | "grassknot" => {
            let kg = reg.species.get(defender.species.as_str()).map(|s| s.weightkg).unwrap_or(0.0);
            return Some(weight_bp(kg));
        }
        "heavyslam" | "heatcrash" => {
            let a = reg.species.get(attacker.species.as_str()).map(|s| s.weightkg).unwrap_or(0.0);
            let d = reg.species.get(defender.species.as_str()).map(|s| s.weightkg).unwrap_or(0.0);
            return Some(relative_weight_bp(a, d));
        }
        "electroball" => {
            let atk = attacker.stat("spe", false);
            let def = defender.stat("spe", false).max(1);
            let ratio = (atk / def).min(4) as usize;
            return Some([40, 60, 80, 120, 150][ratio]);
        }
        "gyroball" => {
            let atk = attacker.stat("spe", false).max(1);
            let power = (25 * defender.stat("spe", false)) / atk + 1;
            return Some(power.min(150));
        }
        "lastrespects" => return Some(50 + 50 * ctx.side_total_fainted),
        "ragefist" => return Some((50 + 50 * ctx.times_attacked).min(350)),
        "eruption" | "waterspout" => {
            let bp = (declared as f64 * attacker.hp as f64 / attacker.maxhp.max(1) as f64).trunc();
            return Some((bp as i64).max(1));
        }
        "hardpress" => {
            let bp = (100.0 * defender.hp as f64 / defender.maxhp.max(1) as f64).trunc();
            return Some((bp as i64).max(1));
        }
        "flail" | "reversal" => {
            let ratio =
                (48.0 * attacker.hp as f64 / attacker.maxhp.max(1) as f64).trunc() as i64;
            return Some(if ratio < 2 {
                200
            } else if ratio < 5 {
                150
            } else if ratio < 10 {
                100
            } else if ratio < 17 {
                80
            } else if ratio < 33 {
                40
            } else {
                20
            });
        }
        "acrobatics" => {
            return Some(if attacker.item.is_none() { declared * 2 } else { declared })
        }
        "assurance" => {
            return Some(if ctx.target_hurt_this_turn { declared * 2 } else { declared })
        }
        "avalanche" => {
            return Some(if ctx.damaged_by_target { declared * 2 } else { declared })
        }
        "storedpower" | "powertrip" => {
            let positive: i64 = attacker.boosts.iter().filter(|v| **v > 0).map(|v| *v as i64).sum();
            return Some(20 + 20 * positive);
        }
        "weatherball" => {
            let doubled = weather_ball_type(ctx.weather_name()).is_some();
            return Some(if doubled { declared * 2 } else { declared });
        }
        "terrainpulse" => {
            let doubled = crate::airborne::terrain_under(attacker, ctx).is_some();
            return Some(if doubled { declared * 2 } else { declared });
        }
        "risingvoltage" => {
            return Some(if ctx.terrain_name() == Some("electricterrain") {
                declared * 2
            } else {
                declared
            })
        }
        "expandingforce" => return Some(declared),
        "tripleaxel" => return Some(20 * ctx.hit_index.max(1)),
        "triplekick" => return Some(10 * ctx.hit_index.max(1)),
        _ => {}
    }

    let conditional = match move_id {
        "stompingtantrum" | "temperflare" => Some(ctx.previous_move_failed),
        "payback" => Some(ctx.moving_last),
        "round" => Some(ctx.ally_used_same_move),
        _ => None,
    };
    if let Some(holds) = conditional {
        return Some(if holds { declared * 2 } else { declared });
    }

    if matches!(move_id, "hex" | "infernalparade" | "barbbarrage" | "venoshock") {
        let statused = if move_id == "venoshock" {
            matches!(defender.status.as_ref().map(|s| s.as_str()), Some("psn") | Some("tox"))
        } else {
            defender.status.is_some()
        };
        return Some(if statused { declared * 2 } else { declared });
    }

    if declared > 0 {
        return Some(declared);
    }
    None
}

/// A move's own `onBasePower` chain entry, as (label, numerator, denominator). At most
/// one ever applies, so it is an `Option` rather than a `Vec` (IKA-101).
pub fn base_power_modifiers(
    reg: &Reg,
    move_id: &str,
    attacker: &Battler,
    defender: &Battler,
    ctx: &MoveContext,
) -> Option<(&'static str, f64, f64)> {
    if move_id == "facade"
        && attacker.status.is_some()
        && attacker.status.map(|s| s.as_str() != "slp").unwrap_or(false)
    {
        return Some(("facade", 2.0, 1.0));
    }
    if move_id == "knockoff"
        && reg.item_is_removable(
            defender.species.as_str(),
            defender.item.as_ref().map(|i| i.as_str()),
        )
    {
        return Some(("knockoff", 1.5, 1.0));
    }
    if matches!(move_id, "solarbeam" | "solarblade") && solar_weak(ctx.weather_name()) {
        return Some(("solar", 0.5, 1.0));
    }
    if move_id == "expandingforce"
        && crate::airborne::terrain_under(attacker, ctx) == Some("psychicterrain")
    {
        return Some(("expandingforce", 1.5, 1.0));
    }
    None
}

/// Terrain's `onBasePower`, for a grounded user's matching type. At most one.
///
/// Misty Terrain's Dragon moves and Grassy Terrain's Earthquake, Bulldoze and Magnitude
/// are halved into a grounded *target*, whatever the user stands on (IKA-201).
pub fn terrain_modifiers(
    move_id: &str,
    move_type: &str,
    attacker_grounded: bool,
    terrain: Option<&str>,
    defender_grounded: bool,
) -> Option<(&'static str, f64, f64)> {
    if terrain == Some("mistyterrain") {
        return (move_type == "Dragon" && defender_grounded)
            .then_some(("mistyterrain", 2048.0, 4096.0));
    }
    if terrain == Some("grassyterrain")
        && matches!(move_id, "earthquake" | "bulldoze" | "magnitude")
    {
        return defender_grounded.then_some(("grassyterrain", 2048.0, 4096.0));
    }
    if !attacker_grounded {
        return None;
    }
    match (terrain, move_type) {
        (Some("electricterrain"), "Electric") => Some(("electricterrain", 5325.0, 4096.0)),
        (Some("grassyterrain"), "Grass") => Some(("grassyterrain", 5325.0, 4096.0)),
        (Some("psychicterrain"), "Psychic") => Some(("psychicterrain", 5325.0, 4096.0)),
        _ => None,
    }
}

File: src/pokeuraou/moveinfo.py
"""Move-specific mechanics: variable base power, type changes, effectiveness overrides.

Showdown implements these as JavaScript callbacks (``basePowerCallback``, ``onModifyType``,
``onEffectiveness``), which the regulation dump cannot serialize -- it only records that
they exist. This module reimplements the ones that occur in the metagame, transcribed from
``vendor/pokemon-showdown/data/moves.ts`` at the pinned commit.

Anything not implemented here is *reported*, not guessed: :func:`base_power` returns
``None`` for a move whose base power it cannot determine, and the caller surfaces that as
an unmodelled effect rather than printing a number with no basis.
"""

from __future__ import annotations

from dataclasses import dataclass

import numpy as np

from .battler import Battler
from .effects import item_is_removable
from .regulation import Regulation

#: Weight thresholds shared by Low Kick and Grass Knot, in hectograms as Showdown stores
#: them (``weightkg`` here is kilograms, so the thresholds are divided by 10).
_WEIGHT_BP: tuple[tuple[float, int], ...] = (
    (200.0, 120),
    (100.0, 100),
    (50.0, 80),
    (25.0, 60),
    (10.0, 40),
)

#: Ratio thresholds for Heavy Slam / Heat Crash.
_RELATIVE_WEIGHT_BP: tuple[tuple[float, int], ...] = ((5.0, 120), (4.0, 100), (3.0, 80), (2.0, 60))

#: Weather that halves Solar Beam / Solar Blade.
_SOLAR_WEAK_WEATHER = frozenset(
    {"raindance", "primordialsea", "sandstorm", "hail", "snowscape", "snow"}
)

#: Weather Ball's type and the doubling that comes with it.
_WEATHER_BALL_TYPE: dict[str, str] = {
    "sunnyday": "Fire",
    "desolateland": "Fire",
    "raindance": "Water",
    "primordialsea": "Water",
    "sandstorm": "Rock",
    "hail": "Ice",
    "snowscape": "Ice",
    "snow": "Ice",
}

_TERRAIN_PULSE_TYPE: dict[str, str] = {
    "electricterrain": "Electric",
    "grassyterrain": "Grass",
    "mistyterrain": "Fairy",
    "psychicterrain": "Psychic",
}


@dataclass(slots=True)
class MoveContext:
    """Per-turn facts a variable-base-power move may need.

    These are things the position does not carry as a matter of course, so the resolver
    supplies them explicitly. A field left at its default is stated, not assumed: for
    instance ``side_total_fainted`` defaults to 0, so Last Respects computed without it is
    computed at its floor -- which is why :func:`base_power` reports rather than silently
    using the default when the value matters.
    """

    weather: str | None = None
    terrain: str | None = None
    #: Fainted count on the attacker's side, for Last Respects.
    side_total_fainted: int = 0
    #: Times the attacker has been hit, for Rage Fist.
    times_attacked: int = 0
    #: Whether the target already took damage this turn, for Assurance.
    target_hurt_this_turn: bool = False
    #: Whether the target damaged the attacker earlier this turn, for Avalanche.
    damaged_by_target: bool = False
    #: Whether the attacker's previous move failed, for Stomping Tantrum and Temper Flare.
    previous_move_failed: bool = False
    #: Which hit of a multi-hit move this is, counting from 1. Triple Axel and Triple Kick
    #: scale their power with it.
    hit_index: int = 1
    #: Whether the attacker moves after the target this turn, for Payback.
    moving_last: bool = False
    #: Whether an ally already used the same move this turn, for Round.
    ally_used_same_move: bool = False


#: Moves whose callback returns the declared base power unless a condition holds. The
#: condition comes from :class:`MoveContext`; with the context left at its defaults these
#: fall back to the declared value, which is the common case rather than a guess.
CONDITIONAL_DOUBLING_BP: dict[str, str] = {
    "stompingtantrum": "previous_move_failed",
    "temperflare": "previous_move_failed",
    "payback": "moving_last",
    "round": "ally_used_same_move",
}

#: Hex and friends double against a statused target, which the position always carries.
STATUS_DOUBLING_BP = frozenset({"hex", "infernalparade", "barbbarrage", "venoshock"})

#: Multi-hit moves whose power grows with the hit number: ``base * move.hit``.
ESCALATING_MULTIHIT_BP: dict[str, int] = {
    "tripleaxel": 20,
    "triplekick": 10,
}

#: Moves whose base power this module computes from state the resolver can supply.
IMPLEMENTED_VARIABLE_BP = frozenset(
    {
        "lowkick", "grassknot", "heavyslam", "heatcrash", "electroball", "gyroball",
        "lastrespects", "ragefist", "eruption", "waterspout", "acrobatics", "assurance",
        "avalanche", "storedpower", "powertrip", "weatherball", "terrainpulse",
        "risingvoltage", "expandingforce", "hardpress", "flail", "reversal",
    }
)

#: Every move whose base power this module can produce.
HANDLED_VARIABLE_BP = (
    IMPLEMENTED_VARIABLE_BP
    | frozenset(CONDITIONAL_DOUBLING_BP)
    | STATUS_DOUBLING_BP
    | frozenset(ESCALATING_MULTIHIT_BP)
)

#: Variable-base-power moves this module does NOT compute. Listed explicitly so the
#: coverage report can name them instead of them failing silently.
KNOWN_UNIMPLEMENTED_VARIABLE_BP = frozenset(
    {
        "beatup", "ficklebeam", "lashout", "mistyexplosion", "gravapple", "shellsidearm",
        "boltbeak", "fishiousrend", "punishment", "trumpcard", "wringout", "crushgrip",
        "spitup", "naturalgift", "fling", "brine", "retaliate", "smellingsalts",
        "wakeupslap", "watershuriken",
    }
)

#: Moves that deal a fraction of the target's current HP instead of rolling damage
#: (Showdown's ``damageCallback``). The value is the divisor.
FRACTIONAL_HP_MOVES: dict[str, int] = {
    "superfang": 2,
    "naturesmadness": 2,
    "ruination": 2,
}

#: Moves whose damage equals the user's remaining HP.
USER_HP_DAMAGE_MOVES = frozenset({"finalgambit"})

#: Moves that bring the target down to the user's HP (``damageCallback``).
ENDEAVOR_MOVES = frozenset({"endeavor"})

#: Damage-callback moves that depend on what hit the user this turn, which a single-turn
#: calculation cannot know on its own.
COUNTER_MOVES = frozenset({"counter", "mirrorcoat", "metalburst", "comeuppance"})


def fixed_damage(
    move_id: str, attacker: Battler, defender: Battler
) -> np.ndarray | None:
    """Damage for moves that bypass the damage formula, or None if not such a move.

    Showdown clamps each of these to at least 1.
    """
    divisor = FRACTIONAL_HP_MOVES.get(move_id)
    if divisor is not None:
        return np.maximum(defender.hp // divisor, 1)
    if move_id in USER_HP_DAMAGE_MOVES:
        return np.maximum(attacker.hp, 1)
    if move_id in ENDEAVOR_MOVES:
        return np.maximum(defender.hp - attacker.hp, 0)
    return None


def _weight_bp(weight_kg: float) -> int:
    for threshold, bp in _WEIGHT_BP:
        if weight_kg >= threshold:
            return bp
    return 20


def _relative_weight_bp(attacker_kg: float, target_kg: float) -> int:
    if target_kg <= 0:
        return 120
    for ratio, bp in _RELATIVE_WEIGHT_BP:
        if attacker_kg >= target_kg * ratio:
            return bp
    return 40


def effective_type(
    reg: Regulation, move_id: str, attacker: Battler, ctx: MoveContext
) -> str | None:
    """Move type after a move's own ``onModifyType``, or None if unchanged.

    Ability-driven type changes (the ``-ate`` abilities, Liquid Voice) are handled in
    ``damage.effective_move_type``; this covers the move's own rules.
    """
    del reg
    if move_id == "weatherball":
        return _WEATHER_BALL_TYPE.get(ctx.weather or "")
    if move_id == "terrainpulse":
        return _TERRAIN_PULSE_TYPE.get(ctx.terrain or "")
    if move_id == "aurawheel":
        # Morpeko-Hangry's form makes it Dark; the base form stays Electric.
        return "Dark" if attacker.species == "morpekohangry" else None
    if move_id == "ragingbull":
        # Tauros-Paldea's three breeds give Raging Bull three types.
        return {
            "taurospaldeacombat": "Fighting",
            "taurospaldeablaze": "Fire",
            "taurospaldeaaqua": "Water",
        }.get(attacker.species)
    return None


def effectiveness_override(
    reg: Regulation, move_id: str, move_type: str, defender_types: tuple[str, ...]
) -> tuple[float, int] | None:
    """A move's ``onEffectiveness``, as (multiplier, Showdown's typeMod).

    Freeze-Dry treats Water as super effective; Flying Press adds Flying's effectiveness
    on top of Fighting's. Both are transcribed rather than approximated because they swing
    damage by a factor of four.
    """
    if move_id == "freezedry":
        mult = 1.0
        steps = 0
        row = reg.typechart.get(move_type, {})
        for t in defender_types:
            if t == "Water":
                mult *= 2.0
                steps += 1
                continue
            m = row.get(t, 1.0)
            mult *= m
            steps += 1 if m == 2 else (-1 if m == 0.5 else 0)
        return mult, max(-6, min(6, steps))

    if move_id == "flyingpress":
        mult = 1.0
        steps = 0
        fighting = reg.typechart.get("Fighting", {})
        flying = reg.typechart.get("Flying", {})
        for t in defender_types:
            for row in (fighting, flying):
                m = row.get(t, 1.0)
                mult *= m
                steps += 1 if m == 2 else (-1 if m == 0.5 else 0)
        return mult, max(-6, min(6, steps))

    return None


def base_power(
    reg: Regulation,
    move_id: str,
    attacker: Battler,
    defender: Battler,
    ctx: MoveContext,
) -> np.ndarray | None:
    """Base power for one hit, or None when this module cannot determine it.

    Returns an array because some formulas depend on stats or HP, which vary across
    belief particles (Electro Ball compares Speeds; Eruption scales with the attacker's
    HP fraction).
    """
    move = reg.moves[move_id]
    n = max(attacker.n, defender.n)
    declared = move.base_power

    def const(value: float) -> np.ndarray:
        return np.full(n, int(value), dtype=np.int64)

    if move_id in ("lowkick", "grassknot"):
        species = reg.species[defender.species]
        return const(_weight_bp(species.weightkg))

    if move_id in ("heavyslam", "heatcrash"):
        return const(
            _relative_weight_bp(
                reg.species[attacker.species].weightkg, reg.species[defender.species].weightkg
            )
        )

    if move_id == "electroball":
        atk_spe = attacker.stat("spe")
        def_spe = np.maximum(defender.stat("spe"), 1)
        ratio = np.minimum(atk_spe // def_spe, 4)
        table = np.array([40, 60, 80, 120, 150], dtype=np.int64)
        return table[ratio]

    if move_id == "gyroball":
        atk_spe = np.maximum(attacker.stat("spe"), 1)
        power = (25 * defender.stat("spe")) // atk_spe + 1
        return np.minimum(power, 150)

    if move_id == "lastrespects":
        return const(50 + 50 * ctx.side_total_fainted)

    if move_id == "ragefist":
        return const(min(350, 50 + 50 * ctx.times_attacked))

    if move_id in ("eruption", "waterspout"):
        # trunc(declared * hp / maxhp), and Showdown clamps to at least 1.
        bp = np.trunc(declared * attacker.hp / np.maximum(attacker.maxhp, 1)).astype(np.int64)
        return np.maximum(bp, 1)

    if move_id == "hardpress":
        bp = np.trunc(100 * defender.hp / np.maximum(defender.maxhp, 1)).astype(np.int64)
        return np.maximum(bp, 1)

    if move_id in ("flail", "reversal"):
        ratio = np.trunc(48 * attacker.hp / np.maximum(attacker.maxhp, 1)).astype(np.int64)
        bp = np.select(
            [ratio < 2, ratio < 5, ratio < 10, ratio < 17, ratio < 33],
            [200, 150, 100, 80, 40],
            default=20,
        ).astype(np.int64)
        return bp

    if move_id == "acrobatics":
        return const(declared * 2 if attacker.item is None else declared)

    if move_id == "assurance":
        return const(declared * 2 if ctx.target_hurt_this_turn else declared)

    if move_id == "avalanche":
        return const(declared * 2 if ctx.damaged_by_target else declared)

    if move_id in ("storedpower", "powertrip"):
        positive = sum(v for v in attacker.boosts.values() if v > 0)
        return const(20 + 20 * positive)

    if move_id == "weatherball":
        return const(declared * 2 if ctx.weather in _WEATHER_BALL_TYPE else declared)

    if move_id == "terrainpulse":
        return const(declared * 2 if ctx.terrain in _TERRAIN_PULSE_TYPE else declared)

    if move_id == "risingvoltage":
        return const(declared * 2 if ctx.terrain == "electricterrain" else declared)

    if move_id == "expandingforce":
        return const(declared)

    escalating = ESCALATING_MULTIHIT_BP.get(move_id)
    if escalating is not None:
        return const(escalating * max(1, ctx.hit_index))

    condition = CONDITIONAL_DOUBLING_BP.get(move_id)
    if condition is not None:
        return const(declared * 2 if getattr(ctx, condition) else declared)

    if move_id in STATUS_DOUBLING_BP:
        statused = defender.status is not None
        if move_id == "venoshock":
            statused = defender.status in ("psn", "tox")
        return const(declared * 2 if statused else declared)

    if declared > 0:
        return const(declared)

    return None


def base_power_is_approximate(move_id: str) -> bool:
    """Whether the value :func:`base_power` returns for this move is a fallback.

    A move in :data:`KNOWN_UNIMPLEMENTED_VARIABLE_BP` still has a declared base power, so
    the calculation produces *a* number. Returning it without saying so is the silent
    wrong answer the reporting split exists to catch, so the caller flags it instead.
    """
    return move_id in KNOWN_UNIMPLEMENTED_VARIABLE_BP


def base_power_modifiers(
    reg: Regulation, move_id: str, attacker: Battler, defender: Battler, ctx: MoveContext
) -> tuple[tuple[str, float, float], ...]:
    """A move's own ``onBasePower`` chain entries, as (label, numerator, denominator)."""
    out: list[tuple[str, float, float]] = []

    if move_id == "facade" and attacker.status is not None and attacker.status != "slp":
        out.append(("facade", 2.0, 1.0))

    elif move_id == "knockoff" and item_is_removable(
        reg, defender.species, defender.item
    ):
        out.append(("knockoff", 1.5, 1.0))

    elif move_id in ("solarbeam", "solarblade") and ctx.weather in _SOLAR_WEAK_WEATHER:
        out.append((move_id, 0.5, 1.0))

    elif move_id == "expandingforce" and ctx.terrain == "psychicterrain":
        out.append(("expandingforce", 1.5, 1.0))

    return tuple(out)


#: Grassy Terrain's `weakenedMoves`, halved into a grounded target.
GRASSY_WEAKENED_MOVES = frozenset({"earthquake", "bulldoze", "magnitude"})


def terrain_modifiers(
    reg: Regulation,
    move_id: str,
    move_type: str,
    attacker_grounded: bool,
    terrain: str | None,
    defender_grounded: bool = True,
) -> tuple[tuple[str, float, float], ...]:
    """Terrain's ``onBasePower``: a grounded user's matching type is boosted; Misty Terrain's
    Dragon moves and Grassy Terrain's Earthquake, Bulldoze and Magnitude are halved into a
    grounded *target*, whatever the user stands on (IKA-201)."""
    del reg
    if terrain == "mistyterrain":
        return (("mistyterrain", 2048, 4096),) if move_type == "Dragon" and defender_grounded else ()
    if terrain == "grassyterrain" and move_id in GRASSY_WEAKENED_MOVES:
        return (("grassyterrain", 2048, 4096),) if defender_grounded else ()
    if terrain is None or not attacker_grounded:
        return ()
    if terrain == "electricterrain" and move_type == "Electric":
        return (("electricterrain", 5325, 4096),)
    if terrain == "grassyterrain" and move_type == "Grass":
        return (("grassyterrain", 5325, 4096),)
    if terrain == "psychicterrain" and move_type == "Psychic":
        return (("psychicterrain", 5325, 4096),)
    return ()
